sql/analyzer/hoist_select_exists.go

```diff
@@ -40,140 +40,130 @@ func hoistSelectExists(
 		if !ok {
 			return n, transform.SameTree, nil
 		}
-
-		s := pluckCorrelatedExistsSubquery(f, len(f.Schema())+len(scope.Schema()))
-		if s == nil {
-			return n, transform.SameTree, nil
-		}
-		if s.right == nil || s.left == nil {
-			panic("unexpected empty scope")
-		}
-
-		if p, ok := s.right.(*plan.Project); ok {
-			s.right = p.Child
-		}
-
-		if gb, ok := s.right.(*plan.GroupBy); ok {
-			s.right = gb.Child
-		}
-
-		outerFilters, _, err := FixFieldIndexesOnExpressions(ctx, scope, a, append(s.left.Schema(), s.right.Schema()...), s.outerFilters...)
-		if err != nil {
-			return n, transform.SameTree, err
-		}
-
-		switch s.joinType {
-		case plan.JoinTypeAnti:
-			return plan.NewAntiJoin(s.left, s.right, expression.JoinAnd(outerFilters...)), transform.NewTree, nil
-		case plan.JoinTypeSemi:
-			return plan.NewSemiJoin(s.left, s.right, expression.JoinAnd(outerFilters...)), transform.NewTree, nil
-		default:
-			panic("expected JoinTypeSemi or JoinTypeAnti")
-		}
+		return hoistExistSubqueries(scope, a, f, len(f.Schema())+len(scope.Schema()))
 	})
 }
 
-type hoistExistsSubquery struct {
-	left         sql.Node
-	right        sql.Node
-	outerFilters []sql.Expression
-	joinType     plan.JoinType
+// simplifyPartialJoinParents discards nodes that will not affect an existence check.
+func simplifyPartialJoinParents(n sql.Node) sql.Node {
+	ret := n
+	for {
+		switch n := ret.(type) {
+		case *plan.Project, *plan.GroupBy, *plan.Limit, *plan.Sort, *plan.Distinct, *plan.TopN:
+			ret = n.Children()[0]
+		case *plan.Filter:
+			panic("unhandled filter")
+		default:
+			return ret
+		}
+	}
 }
 
-// pluckCorrelatedExistsSubquery scans a filter for [NOT] WHERE EXISTS, and then attempts to
+// hoistExistSubqueries scans a filter for [NOT] WHERE EXISTS, and then attempts to
 // extract the subquery, correlated filters, a modified outer scope (net subquery and filters),
 // and the new target joinType
-func pluckCorrelatedExistsSubquery(filter *plan.Filter, scopeLen int) *hoistExistsSubquery {
-	var decorrelated sql.Node
-	var outerFilters []sql.Expression
-
-	filters := splitConjunction(filter.Expression)
-	var newFilters []sql.Expression
-	var joinType plan.JoinType
-	for _, f := range filters {
+func hoistExistSubqueries(scope *Scope, a *Analyzer, filter *plan.Filter, scopeLen int) (sql.Node, transform.TreeIdentity, error) {
+	ret := filter.Child
+	var retFilters []sql.Expression
+	same := transform.SameTree
+	for _, f := range splitConjunction(filter.Expression) {
+		var joinType plan.JoinType
+		var s *hoistSubquery
 		switch e := f.(type) {
 		case *plan.ExistsSubquery:
-			decorrelated, outerFilters = decorrelateOuterCols(e.Query, scopeLen)
-			if len(outerFilters) == 0 {
-				return nil
+			s = decorrelateOuterCols(e.Query, scopeLen)
+			if s == nil {
+				return filter, transform.SameTree, nil
```

Comment:
nit: is it safe/ possible to leave this (and below) as

```
retFilters = append(retFilters, f)
continue
```

instead to pick up cases where one subexpression can't be decorrelated but another can?

Surely the case on 82 should be something like that at least...

Reply:
yeah sorry this was sloppy, i fixed up the easy ones but i think this will need a rewrite the same way as recursive CTEs did. Second round will have more thorough and stable subquery decorrelation.